gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```
@@ -187,82 +193,81 @@ public void startUpload(InputStream pipeSource) {
-          if (responseObserver.hasTransientError()) {
-            continue;
+      boolean objectFinalized = false;
+      while (!objectFinalized) {
+        InsertObjectRequest insertRequest;
+        if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
+          insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
+          writeOffset += insertRequest.getChecksummedData().getContent().size();
+        } else {
+          ByteString data =
+              ByteString.readFrom(ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE));
+          dataChunkMap.put(writeOffset, data);
+          if (dataChunkMap.size() >= NUMBER_OF_REQUESTS_TO_RETAIN) {
+            dataChunkMap.remove(dataChunkMap.firstKey());
           }
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          throw new IOException(
-              String.format("Resumable upload failed for '%s'", getResourceString()), e);
+          insertRequest = buildInsertRequest(writeOffset, data, false);
+          writeOffset += data.size();
         }
+        requestStreamObserver.onNext(insertRequest);
+        objectFinalized = insertRequest.getFinishWrite();
+
+        if (responseObserver.hasTransientError()
+            || responseObserver.hasNonTransientError()) {
+          requestStreamObserver.onError(
+              responseObserver.hasTransientError()
+                  ? responseObserver.transientError
+                  : responseObserver.nonTransientError);
+          break;
+        } else if (objectFinalized) {
+          requestStreamObserver.onCompleted();
+        }
+      }
 
-        return responseObserver.getResponseOrThrow();
+      try {
+        responseObserver.done.await();
+        if (responseObserver.hasTransientError()) {
+          throw new IOException(responseObserver.transientError);
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
```

Is this intentional to swallow exception here?

---

No, here we should throw runtime exception which is the kind that shouldn't be retried.